### PR TITLE
Introduce RequestGraph to help build skinny AssetGraph

### PR DESCRIPTION
**Pull request by padmaia**
# ↪️ Pull Request

The AssetGraphBuilder now builds two graphs in parallel, the "skinny" `AssetGraph` that now only has Dependencies and Assets (and AssetGroups), and the new `RequestGraph` that keeps track of the state of async requests and invalidates them when necessary. This is beneficial because the `AssetGraph` will only have nodes that the Bundling and Packaging phases care about. We can build it incrementally instead of stripping away unimportant nodes after each build, or keeping unimportant nodes and skipping over them while still sending them over IPC. It also makes the `AssetGraph` much easier to read, especially once I start adding more request nodes for config invalidation.   

## 🚨 Test instructions

`yarn test`/`yarn flow`/`yarn lint`

## ✔️ PR Todo

- [x] Clean up `RequestGraph` nodes, when nodes are removed from `AssetGraph`

### CODE REVIEW

File: packages/core/core/test/AssetGraphBuilder.test.js [hunk 1]
```diff
@@ -55,35 +54,12 @@ describe('AssetGraphBuilder', () => {
...
+      builder.assetGraph.nodes.has(
         new Dependency({
           moduleSpecifier: './module-b',
           env: DEFAULT_ENV
         }).id
       )
     );
   });
-
-  it('throws a BuildAbortError when resolving if signal aborts', async () => {
```
Comment (padmaia) [inline packages/core/core/test/AssetGraphBuilder.test.js]: I've removed the `AbortController` code from `AssetGraphBuilder` because we have plans to move it up to Parcel. I plan on adding that back in a separate PR.

**Comment by padmaia** [comment]
Tests were passing locally 😬 I'll try to take a look when I get home

File: packages/core/core/src/AssetGraph.js [hunk 2]
```diff
@@ -4,60 +4,158 @@ import invariant from 'assert';
...
-//   value: asset
-// });
+import Dependency from './Dependency';
+import Graph, {type GraphOpts} from './Graph';
+import type {CacheEntry, DependencyNode, AssetGraphNode, NodeId} from './types';
+
+type AssetGraphOpts = {|
+  onNodeAdded?: (node: AssetGraphNode) => mixed,
+  ...GraphOpts<AssetGraphNode>
+|};
```
Comment (wbinnssmith) [inline packages/core/core/src/AssetGraph.js]: nit: newline

File: packages/core/core/src/AssetGraph.js [hunk 3]
```diff
@@ -4,60 +4,158 @@ import invariant from 'assert';
...
-//   type: 'asset',
-//   value: asset
-// });
+import Dependency from './Dependency';
+import Graph, {type GraphOpts} from './Graph';
+import type {CacheEntry, DependencyNode, AssetGraphNode, NodeId} from './types';
+
+type AssetGraphOpts = {|
+  onNodeAdded?: (node: AssetGraphNode) => mixed,
+  ...GraphOpts<AssetGraphNode>
```
Comment (wbinnssmith) [inline packages/core/core/src/AssetGraph.js]: nit:  `...` is usually first in these cases for overriding properties. but it behaves the same in this case ¯\\\_(ツ)\_/¯

File: packages/core/core/src/AssetGraph.js [hunk 4]
```diff
@@ -72,7 +170,7 @@ export default class AssetGraph extends Graph<AssetGraphNode> {
   }
 
   getDependencyResolution(dep: IDependency): ?Asset {
-    let depNode = this.getNode(dep.id);
+    let depNode = this.nodes.get(dep.id);
```
Comment (wbinnssmith) [inline packages/core/core/src/AssetGraph.js]: why this change?

File: packages/core/core/src/AssetGraphBuilder.js [hunk 5]
```diff
@@ -1,202 +1,113 @@
...
+
+    this.assetGraph = new AssetGraph({
+      onNodeAdded: node => this.handleNodeAddedToAssetGraph(node)
+    });
+    this.requestGraph = new RequestGraph({
       config,
-      options
+      options,
+      onAssetRequestComplete: (request, result) =>
+        this.handleCompletedAssetRequest(request, result),
```
Comment (wbinnssmith) [inline packages/core/core/src/AssetGraphBuilder.js]: since they're both single lines, can each of these functions just be inlined directly here?

File: packages/core/core/src/AssetGraphBuilder.js [hunk 6]
```diff
@@ -1,202 +1,113 @@
...
+  handleCompletedDepPathRequest(
+    requestNode: DepPathRequestNode,
+    result: AssetRequest | null
+  ) {
+    this.assetGraph.resolveDependency(requestNode.value, result);
   }
 
   isInvalid() {
-    return this.graph.invalidNodes.size > 0;
+    return this.requestGraph.isInvalid();
```
Comment (wbinnssmith) [inline packages/core/core/src/AssetGraphBuilder.js]: 😃

File: packages/core/core/src/Parcel.js [hunk 7]
```diff
@@ -242,7 +242,7 @@ export default class Parcel {
           this.#watchEvents.emit({error: err});
           return;
         }
-
+        // console.log('DETECTED CHANGES', events);
```
Comment (wbinnssmith) [inline packages/core/core/src/Parcel.js]: debug :)

File: packages/core/core/src/RequestGraph.js [hunk 8]
```diff
@@ -0,0 +1,227 @@
...
+      return cacheEntry;
+    } catch (e) {
+      // TODO: add connectedFiles even if it failed so we can try a rebuild if those files change
+      throw e;
+    }
+  }
+
+  async resolvePath(dep: Dependency) {
+    try {
+      // console.log('RESOLVING', request);
```
Comment (wbinnssmith) [inline packages/core/core/src/RequestGraph.js]: debug :)

File: packages/core/core/src/types.js [hunk 9]
```diff
@@ -45,26 +44,22 @@ export type DependencyNode = {|
...
-export type TransformerRequestNode = {|
-  id: string,
-  +type: 'transformer_request',
-  value: TransformerRequest
-|};
 export type AssetGroupNode = {|
   id: string,
   +type: 'asset_group',
-  value: AssetGroup
+  value: AssetRequest
```
Comment (wbinnssmith) [inline packages/core/core/src/types.js]: the value for an `AssetGroupNode` is an `AssetRequest`? If so, maybe a different name or a comment explaining why?

File: packages/core/core/src/types.js [hunk 10]
```diff
@@ -45,26 +44,22 @@ export type DependencyNode = {|
...
-  | FileNode
-  | RootNode
-  | TransformerRequestNode
-  | BundleGroupNode
-  | BundleReferenceNode;
+export type DepPathRequestNode = {|
+  id: string,
+  +type: 'dep_path_request',
+  value: Dependency
+|};
```
Comment (wbinnssmith) [inline packages/core/core/src/types.js]: nit: newline

File: packages/core/integration-tests/test/json-reporter.js [hunk 11]
```diff
@@ -19,6 +19,7 @@ describe('json reporter', () => {
     let i = 0;
     // $FlowFixMe
     console.log = function log(msg) {
+      console.log('MSG', msg);
```
Comment (wbinnssmith) [inline packages/core/integration-tests/test/json-reporter.js]: debug :)

File: packages/core/integration-tests/test/server.js [hunk 12]
```diff
@@ -152,7 +152,7 @@ describe('server', function() {
     subscription = await b.watch();
     await getNextBuild(b);
 
-    await sleep(10000);
+    await sleep(4000);
```
Comment (wbinnssmith) [inline packages/core/integration-tests/test/server.js]: can this be removed entirely? not sure why I didn't before since awaiting `getNextBuild` should take care of this

File: packages/examples/simple/src/foo.js [hunk 13]
```diff
@@ -1,3 +1,5 @@
 export default function() {
   return 'hello!';
 }
+
+//syntax\\error
```
Comment (wbinnssmith) [inline packages/examples/simple/src/foo.js]: debug :)

File: packages/core/core/src/AssetGraph.js (hunk 4 above)
Comment (padmaia) [reply packages/core/core/src/AssetGraph.js]: Something was broken with it, I think because I was prepending the node type to the id, but decided to just stop doing that. Thought I put it back to the way it was, good catch.

File: packages/core/integration-tests/test/server.js (hunk 12 above)
Comment (padmaia) [reply packages/core/integration-tests/test/server.js]: looks like it :)

File: packages/core/core/src/RequestGraph.js [hunk 14]
```diff
@@ -0,0 +1,227 @@
...
+      config,
+      options
+    });
+  }
+
+  async initFarm() {
+    // This expects the worker farm to already be initialized by Parcel prior to calling
+    // AssetGraphBuilder, which avoids needing to pass the options through here.
+    this.farm = await WorkerFarm.getShared();
+    this.runTransform = this.farm.mkhandle('runTransform');
```
Comment (devongovett) [inline packages/core/core/src/RequestGraph.js]: I wonder about having the graph being the one doing the actual resolving and transforming, vs being a data structure...

File: packages/core/core/src/types.js [hunk 15]
```diff
@@ -45,26 +44,26 @@ export type DependencyNode = {|
...
-export type OldAssetGraphNode =
-  | AssetNode
-  | AssetReferenceNode
-  | DependencyNode
-  | FileNode
-  | RootNode
-  | TransformerRequestNode
-  | BundleGroupNode
-  | BundleReferenceNode;
+export type DepPathRequestNode = {|
```
Comment (devongovett) [inline packages/core/core/src/types.js]: why `DepPathRequestNode` and not just `DependencyRequestNode`? what is the path part? maybe I missed something

File: packages/core/core/test/AssetGraph.test.js [hunk 16]
```diff
@@ -27,41 +27,40 @@ const stats = {size: 0, time: 0};
...
+      entries: ['/path/to/index1', '/path/to/index2']
     });
 
-    assert(graph.nodes.has('/'));
+    assert(graph.nodes.has('@@root'));
     assert(
       graph.nodes.has(
         new Dependency({
-          moduleSpecifier: './index1',
+          moduleSpecifier: '/path/to/index1',
```
Comment (devongovett) [inline packages/core/core/test/AssetGraph.test.js]: why did these change to absolute?

File: packages/core/utils/src/PromiseQueue.js [hunk 17]
```diff
@@ -21,8 +21,19 @@ export default class PromiseQueue {
     this._maxConcurrent = opts.maxConcurrent;
   }
 
-  add(fn: () => Promise<mixed>): void {
-    this._queue.push(fn);
+  add<T>(fn: () => Promise<T>): Promise<T> {
+    return new Promise((resolve, reject) => {
```
Comment (devongovett) [inline packages/core/utils/src/PromiseQueue.js]: what's this wrapper for?

Comment (padmaia) [reply packages/core/utils/src/PromiseQueue.js]: So that you have access to the results of the promise returned from the function added to the queue. I'm keeping track of in progress request promises so that if that if a an identical request comes in while one is already in progress, it can use the in progress promise. It's not totally needed now as identical dep and asset requests shouldn't fire at the same time, but config request and dev dep requests will bring that possibility. In my config invalidation spike branch `incompleteNodes` turned into `inProgress` promises so I just decided to bring it in now instead of changing it later.

File: packages/core/core/src/types.js (hunk 15 above)
Comment (padmaia) [reply packages/core/core/src/types.js]: I thought it helped to be more specific about what we are requesting. Instead of `DependencyRequest` and `DevDepRequest` which makes it unclear how they are different, we would have `DepPathRequest` and `DepVersionRequest` which should be a little more intuitive. It may not be the best name at the moment because right now we are getting more than just the path from the resolvers (`AssetRequestRequest`? 😛 ), but I have some thoughts on that. I think ideally the resolvers shouldn't have to know what an `AssetRequest` is. Their job should only be to return a path. The AssetGraphBuilder can take care of adding the dep's env to the follow up `AssetRequest`. As for the side effects, I was thinking that getting the `package.json` for a file about to be transformed would work well as a `ConfigRequest` once I have that piece implemented. So I changed the name in anticipation of those changes.

File: packages/core/core/src/RequestGraph.js (hunk 14 above)
Comment (padmaia) [reply packages/core/core/src/RequestGraph.js]: @wbinnssmith and I were talking about making both the `RequestGraph` and `AssetGraph` have instances of a `Graph`. Mainly we were talking about not exposing low level graph methods to users of `AssetGraph`, which is less of a concern for the `RequestGraph` since it is only used by `AssetGraphBuilder`, but that would take care of any concern of having the data structure do async stuff. I like the idea, but I think we can address that in a future PR.

File: packages/core/core/test/AssetGraph.test.js (hunk 16 above)
Comment (padmaia) [reply packages/core/core/test/AssetGraph.test.js]: I noticed that in the initialization of the `AssetGraph` we are not even instantiating the entry deps with `sourcePath`, seemingly because the entries are all resolved to absolute paths before hand so we don't need the `sourcePath` to know where to resolve from. I guess resolving entries to absolute paths is happening in the CLI though. Should we require users of the programmatic API to pass entries as absolute paths?

Comment (wbinnssmith) [reply packages/core/core/test/AssetGraph.test.js]: > Should we require users of the programmatic API to pass entries as absolute paths?

I think we should follow what we're doing with `cacheDir` and target dist information passed through the public api: just do what Node does and resolve things relative to pwd.

Of course, just as you'd do with `fs.readFile`, etc., you're far better off being certain and passing an absolute path.

Comment (wbinnssmith) [reply packages/core/core/test/AssetGraph.test.js]: FWIW, while exploring options overriding node module resolution in #3071, I found that when you run `node` with an entrypoint, it invokes the resolver with a `null` sourcePath equivalent and an absolute path for the `moduleSpecifier` equivalent.
